internal/controller/k0smotron.io/k0smotroncluster_pvc.go

```diff
@@ -0,0 +1,183 @@
+	if err != nil {
+		return fmt.Errorf("failed to reconcile control plane PVC: %w", err)
+	}
+	err = r.reconcileEtcdPVC(ctx, kmc)
+	if err != nil {
+		return fmt.Errorf("failed to reconcile control plane PVC: %w", err)
+	}
+
+	return nil
+}
+
+func (r *ClusterReconciler) reconcileControlPlanePVC(ctx context.Context, kmc km.Cluster) error {
+	// Do nothing if the persistence type is not PVC
+	if kmc.Spec.Persistence.Type != "pvc" {
+		return nil
+	}
+
+	var sts appsv1.StatefulSet
+	err := r.Get(ctx, client.ObjectKey{Namespace: kmc.Namespace, Name: kmc.GetStatefulSetName()}, &sts)
+	if err != nil {
+		// Do nothing if StatefulSet does not exist yet
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+
+		return fmt.Errorf("failed to get statefulset: %w", err)
+	}
+
+	// Do nothing if the sizes match
+	if kmc.Spec.Persistence.PersistentVolumeClaim.Spec.Resources.Requests.Storage().IsZero() ||
+		kmc.Spec.Persistence.PersistentVolumeClaim.Spec.Resources.Requests.Storage().Cmp(*sts.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage()) == 0 {
+		return nil
+	}
+
+	// Update the PVC size
+	var allowExpansion *bool
+	for i := 0; i < int(kmc.Spec.Replicas); i++ {
+
+		if kmc.Spec.Persistence.PersistentVolumeClaim.Name == "" {
+			kmc.Spec.Persistence.PersistentVolumeClaim.Name = kmc.GetVolumeName()
+		}
+		name := fmt.Sprintf("%s-%s-%d", kmc.Spec.Persistence.PersistentVolumeClaim.Name, kmc.GetStatefulSetName(), i)
+
+		var pvc corev1.PersistentVolumeClaim
+		err := r.Get(ctx, client.ObjectKey{Namespace: kmc.Namespace, Name: name}, &pvc)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				// Do nothing if PVC does not exist yet
+				return nil
+			}
+			return fmt.Errorf("failed to get PVC: %w", err)
+		}
+
+		if allowExpansion == nil {
+			var sc storagev1.StorageClass
+			err = r.Get(ctx, client.ObjectKey{Name: *pvc.Spec.StorageClassName}, &sc)
+			if err != nil {
+				return fmt.Errorf("failed to get StorageClass: %w", err)
+			}
+			allowExpansion = sc.AllowVolumeExpansion
+		}
+
+		if allowExpansion != nil && *allowExpansion {
+			pvc.Spec.Resources.Requests[corev1.ResourceStorage] = kmc.Spec.Persistence.PersistentVolumeClaim.Spec.Resources.Requests[corev1.ResourceStorage]
+			err = r.Update(ctx, &pvc)
+			if err != nil {
+				return fmt.Errorf("failed to update PVC: %w", err)
+			}
+
+			// Remove pod to trigger file system resize
+			err = r.Delete(ctx, &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      fmt.Sprintf("%s-%d", kmc.GetStatefulSetName(), i),
+					Namespace: kmc.Namespace,
+				},
+			}, &client.DeleteOptions{})
+
+			if err != nil {
+				return fmt.Errorf("failed to delete pod for resizing: %w", err)
+			}
+		} else {
+			// Do not check other PVCs if expansion is not allowed
+			break
+		}
+	}
+
+	return r.Delete(ctx, &sts, &client.DeleteOptions{PropagationPolicy: ptr.To(metav1.DeletePropagationOrphan)})
+}
+
+func (r *ClusterReconciler) reconcileEtcdPVC(ctx context.Context, kmc km.Cluster) error {
```

Comment:
This implementation looks identical ti the `reconcilePVC` func, would it be possible to have single resize implementation for both cases? Normally some duplication is ok, but in this case the logic in not really trivial and hence having single implemntation for it would be easier to maintain.

Reply:
Yeah, thanks! Done